src/scifem.cpp

```diff
@@ -207,27 +217,73 @@ transfer_meshtags_to_submesh(
+                                std::numeric_limits<T>::min());
   std::vector<std::int32_t> submesh_indices(num_sum_entities);
   std::iota(submesh_indices.begin(), submesh_indices.end(), 0);
 
-  std::span<const std::int32_t> org_indices = tags.indices();
-  std::span<const T> org_values = tags.values();
+  // Map tag indices to global index
+  std::span<const std::int32_t> tag_indices = tags.indices();
+  auto parent_entity_map = topology->index_map(tag_dim);
+  std::vector<std::int64_t> global_tag_indices(tag_indices.size());
+  parent_entity_map->local_to_global(tag_indices, global_tag_indices);
+
+  // Accumulate global indices across processes
+  dolfinx::la::Vector<std::int64_t> index_mapper(parent_entity_map, 1);
+  index_mapper.set(-1);
+  std::span<std::int64_t> indices = index_mapper.mutable_array();
+  for (std::size_t i = 0; i < global_tag_indices.size(); ++i)
+    indices[tag_indices[i]] = global_tag_indices[i];
+  index_mapper.scatter_rev([](std::int32_t a, std::int32_t b)
+                           { return std::max<std::int32_t>(a, b); });
+  index_mapper.scatter_fwd();
+
+  // Map tag values in a similar way (Allowing negative values)
+  dolfinx::la::Vector<T> values_mapper(parent_entity_map, 1);
+  values_mapper.set(std::numeric_limits<T>::min());
+  std::span<T> values = values_mapper.mutable_array();
+  std::span<const T> tag_values = tags.values();
+  for (std::size_t i = 0; i < tag_values.size(); ++i)
+    values[tag_indices[i]] = tag_values[i];
+  values_mapper.scatter_rev([](T a, T b) { return std::max<T>(a, b); });
+  values_mapper.scatter_fwd();
+
   // For each entity in the tag, find all cells of the submesh connected to this
-  // entity
-  for (std::size_t i = 0; i < org_indices.size(); ++i)
+  // entity. Global to local returns -1 if not on process.
+  std::vector<std::int32_t> local_indices(indices.size());
+  parent_entity_map->global_to_local(indices, local_indices);
+  std::span<const T> parent_values = values_mapper.array();
+
+  if (local_indices.size() != parent_values.size())
+    throw std::runtime_error("Number of indices and values do not match");
+
+  for (std::size_t i = 0; i < local_indices.size(); ++i)
   {
-    auto entity_vertices = e_to_v->links(org_indices[i]);
+    assert(i < parent_values.size());
+    assert(i < local_indices.size());
+
+    auto parent_entity = local_indices[i];
+    auto parent_value = parent_values[i];
+    if (parent_entity == -1)
+      continue;
+    assert(parent_entity < e_to_v->num_nodes());
+
+    auto entity_vertices = e_to_v->links(parent_entity);
     bool entity_found = false;
-    for (auto parent_cell : e_to_sub_cell->links(org_indices[i]))
+    for (auto parent_cell : e_to_sub_cell->links(parent_entity))
     {
       if (entity_found)
         break;
-      std::int32_t sub_cell = parent_entity_to_sub_cell[parent_cell];
-      if (sub_cell != -1)
+
+      if (std::int32_t sub_cell = parent_entity_to_sub_cell[parent_cell];
+          sub_cell > -1)
```

Comment:
Apologise for my lack for C++ proficiency, but I think this line is a bit confusing, and I think it looked more readable the way it was written before. If there are good reasons for writing it this way that I don't understand (i.e scoping reasons) than I would really like to know :)

Reply:
This is very similar to how one would use the [walrus](https://realpython.com/python-walrus-operator/) operator in Python.
It is purely for scoping reasons, we do not need `sub_cell` outside this if statement.